src/ray/raylet/scheduling_resources.cc

```diff
@@ -67,75 +62,50 @@ bool ResourceSet::IsEqual(const ResourceSet &rhs) const {
-  // Subtract the resources and track whether a resource goes below zero.
-  bool oversubscribed = false;
+void ResourceSet::SubtractResourcesStrict(const ResourceSet &other) {
+  // Subtract the resources and delete any if new capacity is zero.
   for (const auto &resource_pair : other.GetResourceMap()) {
     const std::string &resource_label = resource_pair.first;
     const double &resource_capacity = resource_pair.second;
     RAY_CHECK(resource_capacity_.count(resource_label) == 1)
         << "Attempt to acquire unknown resource: " << resource_label;
     resource_capacity_[resource_label] -= resource_capacity;
-    if (resource_capacity_[resource_label] < 0) {
-      oversubscribed = true;
+    // TODO(romilb): Double precision subtraction may sometimes be less than zero by a
+    // small epsilon - need to fix.
+    RAY_CHECK(resource_capacity_[resource_label] >=
+              0 - std::numeric_limits<double>::epsilon())
+        << "Capacity of resource " << resource_label << " after subtraction is negative ("
+        << resource_capacity_[resource_label] << ")."
+        << " Debug: resource_capacity_:" << ToString() << ", other: " << other.ToString();
+    if (resource_capacity_[resource_label] == 0) {
+      resource_capacity_.erase(resource_label);
     }
   }
-  return !oversubscribed;
-}
-
-// Perform a left join.
-bool ResourceSet::AddResourcesStrict(const ResourceSet &other) {
-  // Return failure if attempting to perform vector addition with unknown labels.
-  for (const auto &resource_pair : other.GetResourceMap()) {
-    const std::string &resource_label = resource_pair.first;
-    const double &resource_capacity = resource_pair.second;
-    RAY_CHECK(resource_capacity_.count(resource_label) != 0);
-    resource_capacity_[resource_label] += resource_capacity;
-  }
-  return true;
 }
 
 // Perform an outer join.
 void ResourceSet::AddResources(const ResourceSet &other) {
   for (const auto &resource_pair : other.GetResourceMap()) {
     const std::string &resource_label = resource_pair.first;
     const double &resource_capacity = resource_pair.second;
-    if (resource_capacity_.count(resource_label) == 0) {
-      // Add the new label if not found.
-      RAY_CHECK(AddResource(resource_label, resource_capacity));
-    } else {
-      // Increment the resource by its capacity.
-      resource_capacity_[resource_label] += resource_capacity;
-    }
+    resource_capacity_[resource_label] += resource_capacity;
   }
 }
 
-bool ResourceSet::GetResource(const std::string &resource_name, double *value) const {
-  if (!value) {
-    return false;
-  }
+double ResourceSet::GetResource(const std::string &resource_name) const {
   if (resource_capacity_.count(resource_name) == 0) {
-    *value = std::nan("");
-    return false;
+    return 0;
   }
-  *value = resource_capacity_.at(resource_name);
-  return true;
+  double capacity = resource_capacity_.at(resource_name);
+  RAY_CHECK(capacity > 0);
```

Comment:
Should we checking for float precision errors here? Something like `RAY_CHECK(capacity > 0 - std::numeric_limits<double>::epsilon())`

Reply:
I think either way is fine. If it's not causing any issues right now, then we can leave it. @williamma12 is fixing this properly in a separate PR.